synapse/lib/lmdb.py

```diff
@@ -0,0 +1,361 @@
+        if not noindex:
+            self.burs.put(penc + b'\x00' + lval, buid)
+
+        return True
+
+    def put(self, items):
+        '''
+        Put a list of (buid, ((penc, lval, flags),...)) tuples.
+
+        Yields:
+            (buid, <props>) edits (only yields changes).
+        '''
+        for buid, props in items:
+
+            edits = []
+
+            for penc, lval, flags in props:
+
+                if self.set(buid, penc, lval, flags=flags):
+                    edits.append((penc, lval, flags))
+
+            if edits:
+                yield (0, (buid, edits))
+
+class PropStor:
+
+    def __init__(self, lenv, name=b'stor', edits=False):
+
+        self.lenv = lenv
+        self.edits = None
+
+        #self.tags = self.lenv.open_db(b'tags')                             # <tag>00<form>=<init><tick><tock>
+        self.props = self.lenv.open_db(name + b':props', dupsort=True)      # <buid><pkey>=<pval>
+        self.byprop = self.lenv.open_db(name + b':byprop', dupsort=True)    # <pkey>00<pval> = <buid>
+
+    def getPropSetr(self, xact):
+        return PropSetr(self, xact)
+
+    def has(self, xact, penc, byts):
+        with xact.cursor(db=self.byprop) as burs:
+            return burs.set_key(penc + b'\x00' + byts)
+
+    def pref(self, xact, penc, byts):
+        '''
+        Perform a prefix search and yield (buid, penc, pval) rows.
+        '''
+        bkey = penc + b'\x00' + byts
+
+        blen = len(bkey)
+        with xact.cursor(db=self.byprop) as burs:
+
+            if not burs.set_range(bkey):
+                return
+
+            for lkey, buid in burs.iternext():
+
+                if lkey[:blen] != bkey:
+                    return
+
+                renc, rval = lkey.split(b'\x00', 1)
+                yield (buid, renc, rval)
+
+    def range(self, xact, penc, bval, nval):
+        '''
+        Perform a range search and yield (buid, penc, pval) rows.
+        '''
+        bkey = penc + b'\x00' + bval
+        nkey = penc + b'\x00' + nval
+
+        blen = len(bkey)
+        with xact.cursor(db=self.byprop) as burs:
+
+            if not burs.set_range(bkey):
+                return
+
+            for lkey, buid in burs.iternext():
+
+                if lkey >= nkey:
+                    break
+
+                renc, rval = lkey.split(b'\x00', 1)
+                yield (buid, renc, rval)
+
+    def recs(self, xact, rows):
+        '''
+        Yield full (buid, (props..)) records from rows.
+        '''
+        with xact.cursor(db=self.props) as purs:
+
+            for buid, penc, pval in rows:
+
+                # props: <buid><prop>=<valu>
+                if not purs.set_range(buid):
+                    # yield empty results for iterator alignment..
+                    yield (buid, ())
+                    continue
+
+                props = []
+
+                for lkey, lval in purs.iternext():
+                    props.append((lkey[32:], lval))
```

Comment:
@invisig0th @vEpiphyte - what do we actually want to append here?

Reply:
Exactly that.  :)